### ISSUE TITLE
[BUG] Calling waitForNavigation twice doesn't work

### ISSUE THREAD

**Issue by celeryclub** [microsoft/playwright#1050]
**Context:**
- Playwright Version: 0.11.1
- Operating System: macOS
- Extra: Node 12.14.1

**Code Snippet**
Here's a minimal repro of the behavior I'm seeing.

```javascript
const { chromium, webkit, firefox } = require('playwright');

(async () => {
  const browser = await chromium.launch({ headless: false });
  const page = await browser.newPage();

  await page.goto('https://www.behance.net/search');
  // await page.waitForNavigation({ waitUntil: 'networkidle0' });

  // temporary - to ensure JS bindings are installed so that the history API will always be used for the second navigation
  await page.waitFor(2000);

  await Promise.all([
    page.click('[data-ut="moodboards-tab"]'),
    page.waitForNavigation({ waitUntil: 'networkidle0' }),
  ]);

  await page.screenshot({path: 'repro.png'});

  await browser.close();
})();

```

**Describe the bug**

This is a follow-up to https://github.com/microsoft/playwright/issues/996.

It seems that when I call `waitForNavigation` twice in the same context, the second call doesn't work. In this case the second call is a history API navigation event, so that could be part of the reason too.

When I go to a page without any extra waiting, then click a link to navigate and wait for that navigation, everything works fine. But when I also try to wait for the first navigation, the second wait resolves immediately without the page being fully loaded. I've tried a couple of variations of this, including passing `waitUntil: 'networkidle0' }` into `page.goto` instead of calling `waitForNavigation`, and using a `Promise.all` for the first navigation too. All of these produce the same result.

Perhaps there's something I'm misunderstanding about how `waitForNavigation` works? The docs seem to indicate that history API changes are considered relevant navigation events too.

When I don't `waitForNavigation` as part of the first `page.goto`, the wait for the second navigation works correctly:
![repro1](https://user-images.githubusercontent.com/750809/74675626-ebfbcc00-5181-11ea-9b34-b37b1a20cd02.png)

When I do `waitForNavigation` as part of the first `page.goto`, the wait for the second navigation doesn't work:
![repro2](https://user-images.githubusercontent.com/750809/74675674-0635aa00-5182-11ea-8113-2905d65270fc.png)

Thanks for taking a look!

**Comment by JoelEinbinder** [comment]
To start `waitForNavigation` waits for the next navigation, and then for some state after it if you specify a waitUntil. You are correct that it waits for either normal navigations, or usages of the history API. There is also an implied waitForNavigation inside `page.goto`.

The issue is that the page calls `history.replaceState` multiple times, I think to change the title. You can see the navigations with this snippet:

```js
  page.on('framenavigated', frame =>  {
    if (frame === page.mainFrame())
      console.log(frame.url());
  });
```

So depending on how many have happened, you are waiting for the wrong navigation. Additionally `networkidle0` waits for network requests to finish, but javascript can still run after that. For example to populate images. So you should really be waiting for something specific to your website, possibly using `waitForFunction`. Otherwise it will always be flaky.

**Comment by celeryclub** [comment]
@JoelEinbinder thanks so much for taking the time to look into this. I really appreciate it. I understand what you're saying about `replaceState`, but I'm not sure why this function being called multiple times would cause the behavior I'm seeing. The second navigation is the problematic one (`page.click('[data-ut="moodboards-tab"]')`) and that's happening after all of the extra `replaceState` calls have finished.

To put it another way, after `page.waitFor(2000)`, the page has fully settled, and yet the navigation that occurs after that point isn't being waited for correctly. Is this the expected behavior? If so, I don't quite follow the logic of it.

Thanks again for your time and for helping me understand this better.

**Comment by arjun27** [comment]
Thanks for this example @celeryclub. Sharing my observations after some investigation in the hope of providing hints (apologies in advance for not having a solution):

If I navigate directly to /search/moodboards (with `page.goto`), I'm seeing multiple `framenavigated` events on the main frame (similar to the navigation to /search). The first such event is (probably) resolving the second `page.waitForNavigation` in your example. This is to say that the extra `replaceState` calls get triggered again when the `moodboards-tab` element is clicked.

This behavior however raises the question why the second waitForNavigation does not respect the `{ waitUntil: "networkidle0" }` option. I would expect the wait to not resolve if there is ongoing network activity (which is happening when the element is clicked).

My suspicion was that there is a timing issue: the first `replaceState` call happens when the network is idle, causing the wait to resolve. To verify this, I tried to monkey-patch `replaceState` to console log it, and verify whether that happens before the first network activity. I also monkey-patched the `pushState` method. On clicking the `moodboards-tab` element, I could see the `pushState` call, but I don't see any calls to `replaceState`. Which makes me question whether `replaceState` is being called at all–and if it isn't, the `framenavigated` event should not be fired?

Code used to monkey-patch

```js
const pushState = history.pushState;
history.pushState = function () {
    pushState.apply(history, arguments);
    console.log('pushState', arguments);
};
```

**Comment by celeryclub** [comment]
Thanks so much for looking into this @arjun27. This does provide a bit more context, but like you said, doesn't seem to explain it fully. I really appreciate you taking some time to dig into it though. Hope we can get to the bottom of it!

**Comment by celeryclub** [comment]
I did a bit more digging into how the site works, and here's a timeline of what I found:

1. pw navigates to https://www.behance.net/search
2. `replaceState` is called with the route `/search` and a title
3. `replaceState` is called again with the same route and title
4. pw clicks "Moodboards" which causes a navigation to https://www.behance.net/search/moodboards
5. `pushState` is called with the route `/search/moodboards` but no title

When I use pw to watch for `framenavigated` events using the snippet above provided by @JoelEinbinder, I see 6 of them.

```sh
https://www.behance.net/search
https://www.behance.net/search
https://www.behance.net/search/moodboards
https://www.behance.net/search/moodboards
https://www.behance.net/search/moodboards
https://www.behance.net/search/moodboards
```

The first 3 of those look correct to me, but I'm not sure where the last 3 are coming from. In this research I only looked for `replaceState` and `pushState` calls. Are there other calls that are included in `framenavigated` that I should also check for?

**Comment by JoelEinbinder** [comment]
Can you run your script with ‘DEBUG=* node yourfile.js’? It will dump a ton of output and we can see what’s causing the extra navigations.

**Comment by celeryclub** [comment]
Wow that really is a ton. :) Here's what I got when I ran that. https://gist.github.com/celeryclub/6260825eb86f647cd8befaba1189c701

**Comment by dgozman** [comment]
I _think_ the issue here is that `pushState` does not reset networkidle detection, and so the second `waitForNavigation` call just resolves immediately after the `pushState`. That sounds like a bug.

**Comment by pavelfeldman** [comment]
@dgozman, is that fixed?

**Comment by dgozman** [comment]
We decided that current behavior is fine. The right way to use "network idle" in this case is to ensure that all requests started **after** the action has been finished. There needs to be some wrapper that listens to `request`, `requestfinished` and `requestfailed` events during the action. Something along these lines:

```js
async function wrapWithNetworkIdle(page, action) {
  let networkSettledCallback;
  const networkSettledPromise = new Promise(f => networkSettledCallback = f);

  let requestCounter = 0;
  const onRequest = () => ++requestCounter;
  const onRequestDone =() => {
    // Let the page handle responses asynchronously (via setTimeout(0)).
    const evaluate = page.evaluate(() => new Promise(f => setTimeout(f, 0)));
    evaluate.catch(e => null).then(() => {
      if (!--requestCounter)
        networkSettledCallback();
    });
  };

  page.on('request', onRequest);
  page.on('requestfinished', onRequestDone);
  page.on('requestfailed', onRequestDone);

  const result = await action();
  await networkSettledPromise;
  
  page.removeListener('request', onRequest);
  page.removeListener('requestfinished', onRequestDone);
  page.removeListener('requestfailed', onRequestDone);

  return result;
}
```

We'll experiment with this later.

**Comment by celeryclub** [comment]
Thanks for the follow-up @dgozman. This makes sense to me. 👍
